### DISCUSSION TITLE
Promise rejections and error thrown are shown in failureReason property instead of error property

### DISCUSSION THREAD

**Discussion by dnyaneshwar2023** [TanStack/query#6271] [Q&A] [answered]
I am trying out react query for the first time and I found that something is behaving differently than as given in the documentation.

React Query Version: 5.0.0 (using with Next.js)

![image](https://github.com/TanStack/query/assets/122769564/f0cc7b84-fe65-4171-8398-83b1729b1a33)

I modified my fetcher function little bit to test both Promise rejection and the throwing error scenario.

Scenario 1:

```JS
const result = useQuery({
    queryKey: ['todos', 1],
    queryFn: async () => {
      if (true) {
        throw new Error('Oh no!')
      }
      if (false) {
        return Promise.reject(new Error('Oh no!'))
      }
  
      return []
    },
  });

  console.log(result)
```

Output: 
![image](https://github.com/TanStack/query/assets/122769564/bcb82041-346a-4174-96d3-3341975944f8)

Scenario 2: (With promise rejection)
```JS
const result = useQuery({
    queryKey: ['todos', 1],
    queryFn: async () => {
      if (false) {
        throw new Error('Oh no!')
      }
      if (true) {
        return Promise.reject(new Error('Oh no! with promise rejection'))
      }
  
      return []
    },
  });

  console.log(result)
```

Output: 
![image](https://github.com/TanStack/query/assets/122769564/35a3ad87-0863-4ee4-a457-090f77e0eeae)

Correct me if I am missing something.

**Comment by TkDodo** [comment, marked as answer]
react-query does 3 retries per default. `failureCount` increases to 3 with the underlying `failureReason` before the query actually goes into error state. if you want instant error state, set `retry: false`

**Reply by dnyaneshwar2023** [reply]
Do you think it needs to be mentioned in the documentation? I am happy to contribute on it.

**Reply by TkDodo** [reply]
this is mentioned in multiple places I think

**Reply by dnyaneshwar2023** [reply]
https://tanstack.com/query/latest/docs/react/guides/query-functions?from=reactQueryV3&original=https%3A%2F%2Ftanstack.com%2Fquery%2Fv3%2Fdocs%2Fguides%2Fquery-functions#handling-and-throwing-errors

I think this part it's not much clear that error property will be updated after all retry attempts are done.
This was the point which I got this for the first time.

**Reply by TkDodo** [reply]
if you want, you can amend that information here: https://tanstack.com/query/latest/docs/react/guides/query-retries

**Reply by dnyaneshwar2023** [reply]
Sure, Thanks for your help!!
